### ISSUE TITLE
DNS update of api.internal.$clustername after rolling-upgrade is failing

### ISSUE THREAD

**Issue by pierreozoux** [kubernetes/kops#3351]
I just performed a rolling upgrade of my cluster, and noticed that I have 4 records with 3 master nodes.
After checking, I realized that there is still the private IP address of one old master.

Here the logs from:

```

I0907 12:11:30.607849       1 dnscontroller.go:611] Update desired state: pod/kube-system/kube-apiserver-ip-10-221-102-187.ec2.internal: [{A api.internal.$clustername 10.221.102.187 false}]
I0907 12:11:31.929287       1 dnscontroller.go:256] Using default TTL of 1m0s
I0907 12:11:31.929320       1 dnscontroller.go:421] Querying all dnsprovider records for zone "$clustername."
I0907 12:11:32.039869       1 dnscontroller.go:567] will replace existing dns record A api.internal.$clustername.
I0907 12:11:32.039891       1 dnscontroller.go:571] Adding DNS changes to batch {A api.internal.$clustername.} [10.221.102.187 10.221.106.168 10.221.55.143 10.221.79.203]
I0907 12:11:32.039935       1 dnscontroller.go:301] applying DNS changeset for zone $clustername.::xxx
I0907 12:13:10.195465       1 dnscontroller.go:611] Update desired state: node/ip-10-221-33-64.ec2.internal: []
I0907 12:15:11.169995       1 dnscontroller.go:611] Update desired state: node/ip-10-221-81-221.ec2.internal: []
I0907 12:15:36.905655       1 dnscontroller.go:611] Update desired state: node/ip-10-221-54-17.ec2.internal: [{A node/ip-10-221-54-17.ec2.internal/internal 10.221.54.17 true} {A node/role=node/internal 10.221.54.17 true} {A node/role=node/ ip-10-221-54-17.ec2.internal true} {A node/role=node/ ip-10-221-54-17.ec2.internal true}]
I0907 12:16:20.837139       1 dnscontroller.go:611] Update desired state: node/ip-10-221-89-232.ec2.internal: [{A node/ip-10-221-89-232.ec2.internal/internal 10.221.89.232 true} {A node/role=node/internal 10.221.89.232 true} {A node/role=node/ ip-10-221-89-232.ec2.internal true} {A node/role=node/ ip-10-221-89-232.ec2.internal true}]
I0907 12:17:03.314664       1 dnscontroller.go:611] Update desired state: node/ip-10-221-122-153.ec2.internal: []
I0907 12:18:26.371609       1 dnscontroller.go:611] Update desired state: node/ip-10-221-102-201.ec2.internal: [{A node/ip-10-221-102-201.ec2.internal/internal 10.221.102.201 true} {A node/role=node/internal 10.221.102.201 true} {A node/role=node/ ip-10-221-102-201.ec2.internal true} {A node/role=node/ ip-10-221-102-201.ec2.internal true}]
I0907 12:43:11.778229       1 podcontroller.go:102] pod watch channel closed
W0907 12:43:11.778258       1 podcontroller.go:69] querying without label filter
W0907 12:43:11.778263       1 podcontroller.go:71] querying without field filter
W0907 12:43:11.805712       1 podcontroller.go:84] querying without label filter
W0907 12:43:11.805729       1 podcontroller.go:86] querying without field filter
```

After a restart of this pod, all is alright.

kops version: 1.7.0
k8s version: 1.7.5

Is it normal behavior?

**Comment by chrislovecnm** [comment]
This is a known bug

**Comment by blakebarnett** [comment]
Is there another issue or PR to reference here?

**Comment by chrislovecnm** [comment]
I have seen this reported a couple of times. We have an issue where an old master is hanging around. This is most likely with using dns for masters and not an elb. Have not had a chance to recreate

**Comment by chrislovecnm** [comment]
Old master hanging around means that the dns controller is setting an address for a master that. No longer exists. It may be a race condition with the k8s API where the old master is still registered, I have not looked at the code, or tried to re create.

**Comment by dcowden** [comment]
We are experiencing behavior that might be explained by this problem. When we do a rolling upgrade of the masters, the nodes in the zone matching the masters go NotReady for about 10 minutes.

We tracked the problem down to the fact that kube-dns lookups timeout on the worker during this period.

We changed our api.internal name to point to an elb instead of a round-robin of the master nodes, but this did not address the issue for us.
